File: src/test/java/equidiaryDB/config/ConfigTest.java
package equidiaryDB.config;

import equidiaryDB.GenericTestCase;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import static equidiaryDB.TestConstant.*;
import static org.junit.Assert.assertEquals;

public class ConfigTest extends GenericTestCase {

    @After
    public void tearsDown() throws IOException {
        Files.deleteIfExists(CONFIG_PROPERTIES);
    }

    @Test
    public void nullConfigIfFileDoesntExists() throws IOException {
        givenNoConfigFile();

        Config config = whenCreateConfig();

        thenGetNullConfig(config);
    }

    @Test
    public void portIsInvalid() throws IOException {
        givenConfigProperties(CONFIG_TEST_INVALID_IP);

        Config config = whenCreateConfig();

        thenGetNullConfig(config);
    }


    // GIVEN

    private void givenNoConfigFile() throws IOException {
        Files.deleteIfExists(CONFIG_PROPERTIES);
    }

    private void givenConfigProperties(Path configFile) throws IOException {
        Files.copy(configFile, CONFIG_PROPERTIES);
    }

    // WHEN

    private Config whenCreateConfig() throws IOException {
        return Config.createConfig(CONFIG_PROPERTIES);
    }

    // THEN

    private void thenGetNullConfig(Config config) {
        assertEquals(NullConfig.INSTANCE, config);
    }
}

File: src/test/java/equidiaryDB/config/ConfigTest.kt
package equidiaryDB.config

import equidiaryDB.GenericTestCase
import equidiaryDB.TestConstant
import equidiaryDB.config.Config.Companion.NULL_CONFIG
import equidiaryDB.config.Config.Companion.createConfig
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path

class ConfigTest : GenericTestCase() {
    @AfterEach
    @Throws(IOException::class)
    fun tearsDown() {
        Files.deleteIfExists(TestConstant.CONFIG_PROPERTIES)
    }

    @Test
    @Throws(IOException::class)
    fun nullConfigIfFileDoesntExists() {
        givenNoConfigFile()
        val config = whenCreateConfig()
        thenGetNullConfig(config)
    }

    @Test
    @Throws(IOException::class)
    fun portIsInvalid() {
        givenConfigProperties(TestConstant.CONFIG_TEST_INVALID_IP)
        val config = whenCreateConfig()
        thenGetNullConfig(config)
    }

    // GIVEN
    @Throws(IOException::class)
    private fun givenNoConfigFile() {
        Files.deleteIfExists(TestConstant.CONFIG_PROPERTIES)
    }

    @Throws(IOException::class)
    private fun givenConfigProperties(configFile: Path?) {
        Files.copy(configFile, TestConstant.CONFIG_PROPERTIES)
    }

    // WHEN
    @Throws(IOException::class)
    private fun whenCreateConfig(): Config? {
        return createConfig(TestConstant.CONFIG_PROPERTIES)
    }

    // THEN
    private fun thenGetNullConfig(config: Config?) {
        Assertions.assertEquals(NULL_CONFIG, config)
    }
}
